lib/vdsm/tool/configurators/multipath.py

```diff
@@ -109,3 +106,25 @@ def isconfigured():
 
     sys.stdout.write("multipath requires configuration\n")
     return NO
+
+
+def isconfigured():
+    """
+    Check the multipath daemon configuration.
+    """
+    conf_metadata_ok = _check_mpath_metadata()
+    try:
+        mpathconf.check_local_config()
+    except exception.InvalidConfiguration as e:
+        ufn_enabled_sections = e.context["sections"]
+        for section in ufn_enabled_sections:
+            sys.stdout.write(
+                f"Invalid configuration detected: {e.context['reason']}:\n")
+            sys.stdout.write(f"  {section.name} {{\n")
+            for attr in section.children:
+                sys.stdout.write(f"    {attr.key} {attr.value}\n")
+            sys.stdout.write("  }\n")
+            sys.stdout.write("This configuration is not supported "
+                             "and may lead to storage domain corruption.\n")
```

Comment:
It would be better to use a specific error and format the error message in the exception class __str__(). The we will simplify do there:

    try:
        check config...
    except InvalidMultipathConfiguration as e:
        sys.stdout.write("WARNING: %s", e)

Reply:
I think is good as it is. The exception is a generic container with a bit of context.

`sections` contain the raw contents of the configuration structure, so it will get printed as:
```
Invalid configuration value: 'user_friendly_names' is enabled in multipath configuration {'sections': [Section(name='defaults', children=[Pair(key='key', value='value'), Pair(key='user_friendly_names', value='yes')]), Section(name='device', children=[Pair(key='user_friendly_names', value='yes'), Pair(key='key', value='value')])]}
```

Which is horrible for users. This structure needs to get transformed into a readable format, but the exception class is agnostic to whatever the context dictionary may contain. Then, we either transform it before in the mpathconf module or after, in the configurator. IMO, the configurator is a good place to do it, is where we have full control over what we want to show.

In here I just need to get the first and last lines out of the loop, and it looks good. I indented it wrongly in some update.

Reply:
The exception __str__ can return a string with multiple lines - the same thing you output here. But it is true that it will be harder to control how the output look in the tool.

If you want to format the message here, use more specific exception that has a public
"issues" or "sections" attribute instead of abusing the context dict.

Reply:
Something like this?

```python
class InvalidMPathConfiguration(InvalidConfiguration):
    code = 103
    message = "Invalid configuration value"

    def __init__(self, reason, sections):
        super().__init__(reason)
        self.sections = sections
    
    def __str__(self):
        error_str = f"{self.msg}: {self.context['reason']}:\n"
        for section in self.sections:
            error_str += f"  {section.name} {{\n"
            for attr in section.children:
                error_str += f"    {attr.key} {attr.value}\n"
            error_str += "  }\n"
        error_str += "This configuration is not supported " \
                     "and may lead to storage domain corruption.\n"
        return error_str
```

Reply:
I would use io.StringIO to format, but this is the idea. If you do the formatting
in `__str__` you don't need to provide a public `sections` attribute. If you want to format the error in the tool, using a public attribute is better.

Reply:
StringIO sounds good. I will probably format the error in the _ _str_ _. As we are introducing a new exception class anyway, my argument above about keeping it agnostic to context is not important.

Reply:
At the end what matters is what works best for our needs. The general approach in vdsm to to format errors in a useful way in the exception class. However we sue this only for logging when we care less about the formatting and more about the content. If formatting the error in the tool is easier do it in the tool.

Another option is to avoid the exception and instead return the result of the check to the tool, and format the warning in the tool. Even if we want to fail, we can do this
in the tool instead of the mpathconf module.